[BOL-46] 파일 업로드 (MultiPart) 구현

prk4224 opened the pull request:
* refactor: Interceptor  에 ContenetType 제거

* feat: 파일 업로드 Response 생성

* feat: ImageFileApi 생성

* feat: ImageFileApi 의존성 주입

* refactor: DataSourceModule Provide -> Binds

* feat: 파일 업로드 함수 구현 (DataSource)

* feat: 이미지 타입 생성

* feat: 피일 업로드 로직 구현 및 NewGroupUseCase 생성

* feat: NewGroupViewModel 생송

* feat: NewGroupActivity 생성

* feat: 로그인 화면에서 그룹 생성 화면 으로 토큰 값 보내기

* feat: kakaotestViewModel 생성

* feat: 읽기/쓰기 권한 추가

* feat: Glide 빌드 추가

* feat: 갤러리 접근 시 사용되는 데이터 상수화

* feat: getCursor 구현

* feat: setGroupImage 구현

* feat: getRealPathFromURI 구현 (진짜 파일로 만들어주는 함수)

* feat: getResultLauncher 구현

* feat: 권한 가져오기 & 체크 구현

* feat: generateGallery 함수 구현

* feat:checkedGalleryAccess 함수 구현

* feat: 버튼 클릭 시 갤러리 띄우기 & 파일 업로드 함수 호출

### Reference
- [Jira Ticket](https://yapp22-aos2.atlassian.net/browse/BOL-)

### 참고 사항

P1: 꼭 반영해주세요 (Request changes)
P2: 적극적으로 고려해주세요 (Request changes)
P3: 웬만하면 반영해 주세요 (Comment)
P4: 반영해도 좋고 넘어가도 좋습니다 (Approve)
P5: 그냥 사소한 의견입니다 (Approve)

prk4224 committed:
Feat: 파일 업로드 로직 구현 (#13)

* refactor: Interceptor  에 ContenetType 제거

* feat: 파일 업로드 Response 생성

* feat: ImageFileApi 생성

* feat: ImageFileApi 의존성 주입

* refactor: DataSourceModule Provide -> Binds

* feat: 파일 업로드 함수 구현 (DataSource)

* feat: 이미지 타입 생성

* feat: 피일 업로드 로직 구현 및 NewGroupUseCase 생성

* feat: NewGroupViewModel 생송

* feat: NewGroupActivity 생성

* feat: 로그인 화면에서 그룹 생성 화면 으로 토큰 값 보내기

* feat: kakaotestViewModel 생성

* feat: 읽기/쓰기 권한 추가

* feat: Glide 빌드 추가

* feat: 갤러리 접근 시 사용되는 데이터 상수화

* feat: getCursor 구현

* feat: setGroupImage 구현

* feat: getRealPathFromURI 구현 (진짜 파일로 만들어주는 함수)

* feat: getResultLauncher 구현

* feat: 권한 가져오기 & 체크 구현

* feat: generateGallery 함수 구현

* feat:checkedGalleryAccess 함수 구현

* feat: 버튼 클릭 시 갤러리 띄우기 & 파일 업로드 함수 호출

ckrudals commented on app/src/main/java/com/yapp/bol/app/di/AppInterceptor.kt:
p1 : 현재 이 인터셉터에서 하고 있는 기능이 없는 것 같아요. 그래서 제가 [[BOL-10] accessToken, refreshToken 관리 기능 추가](https://github.com/YAPP-Github/22nd-Android-Team-2-Android/pull/14) pr 를 수행하면서 제거했는데 merge 할 때 되살아 나지 않게 주의해주세요.

ckrudals commented on app/src/main/java/com/yapp/bol/app/di/NetworkModule.kt:
```diff
@@ -70,6 +72,13 @@ object NetworkModule {
+
+    @Singleton
+    fun provideFileUploadApiService(retrofit: Retrofit): ImageFileApi {
+        return retrofit.create(ImageFileApi::class.java)
```
p3 : `ImageFileAPI` 가 여러 화면에서 사용된다면 유지하고, 만약 그룹생성할 때 만사용되면 `GroupAPI` 혹은 `CreateGroupAPI` 로 변경하시는 것은 어떤가요? 

이름만 보고 어디서 사용될지 추측하기가 어렵네요.

ckrudals commented on data/src/main/java/com/yapp/bol/data/datasource/impl/MockDataSourceImpl.kt:
```diff
@@ -1,20 +1,58 @@
+
+       return safeApiCall(remoteErrorEmitter) {
+           imageFileApi.postFileUpload(
+               token = token.convertRequestToken(),
```
p1 : header 에 token 넣어주는 부분은 [BOL-10] accessToken, refreshToken 관리 기능 추가](https://github.com/YAPP-Github/22nd-Android-Team-2-Android/pull/14) 이 pr 에서 인터셉터를 통해 자동으로 넣어주는 기능을 만들었기 때문에 불필요 할 것 같아요.

ckrudals commented on data/src/main/java/com/yapp/bol/data/remote/ImageFileApi.kt:
```diff
@@ -0,0 +1,21 @@
+    suspend fun postFileUpload(
+        @Header("Authorization") token: String,
+        @Part file: MultipartBody.Part,
+        @Part("purpose") purpose: RequestBody,
```
p4 :  `MultipartBody.Part` , `RequestBody` 을 각각 주입하고 있는 있는데 이유가 있나요?

ckrudals commented on data/src/main/java/com/yapp/bol/data/repository/MockRepositoryImpl.kt:
```diff
@@ -18,4 +19,8 @@ class MockRepositoryImpl @Inject constructor(
     }
+
+    override suspend fun postFileUpload(remoteErrorEmitter: RemoteErrorEmitter, token: String, file: File): String {
+        return mockDataSource.postFileUpload(remoteErrorEmitter,token,file)?.url ?: "Null !!!"
```
p2 : url 이 null 일때 emptyString("") 를 반환하는게 의도한 것 같은데, `Null !!!` 라는 String 을 사용하는 이유가 있나요? `.orEmpty()` , `?: ""` 와 같은 형식을 사용하지 않고 특별한 String 을 반환하는 이유가 궁금해요, 이 url 을 받는 쪽에서 `Null !!!` 값이 올 거라고 쉽게 예상하지 못할 것 같아요.

ckrudals commented on domain/src/main/java/com/yapp/bol/domain/usecase/login/NewGroupUseCase.kt:
```diff
@@ -0,0 +1,14 @@
+import java.io.File
+import javax.inject.Inject
+
+class NewGroupUseCase @Inject constructor(
```
p2 : usecase 의 이름은 는 동사 + 명사/대상(선택사항) + usecase  로 짓는게 규칙입니다. 따라서 `CreatNewGroupUsecase` 정도로 이름을 지을 수 있겠네요. [이 가이드의 이름 지정 규칙](https://developer.android.com/topic/architecture/domain-layer?hl=ko#conventions)

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/login/KakaoTestActivity.kt:
```diff
@@ -40,6 +42,13 @@ class KakaoTestActivity : AppCompatActivity() {
         kakaoLogin()
+
+        kakaoTestViewModel.accessToken.observe(this) {
+            if(it == EMPTY_STRING) return@observe
```
p2 : `EMPTY_STRING`를 사용하는 것보단 kotlin collection 함수를 사용하는 것은 어떤가요? 여기서는 `it.isNullOrEmpty()` 로 사용할 수 있겠네요.

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/login/KakaoTestActivity.kt:
정렬이 필요합니다.

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/login/KakaoTestActivity.kt:
```diff
@@ -40,6 +42,13 @@ class KakaoTestActivity : AppCompatActivity() {
+        kakaoTestViewModel.accessToken.observe(this) {
+            if(it == EMPTY_STRING) return@observe
+            val intent = Intent(this, NewGroupActivity::class.java)
+            intent.putExtra(ACCESS_TOKEN,it)
```
p2 : 제가 `accessToken` 을 가져오는 usecase 를 만들었으니, Intent 로 전달하는 것보단 그 화면의 viewModel에서 usecase 를 통해 `subscribe` 하는게 구조적으로 좋아보이네요.

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/login/NewGroupActivity.kt:
```diff
@@ -0,0 +1,115 @@
+        binding = ActivityNewGroupBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+
+        val accessToken = intent.getStringExtra(ACCESS_TOKEN) ?: ""
```
p3: 알고 있을지는 모르겠지만. intent 로 넘어온 value 를 viewModel 에서 `savedStateHandle` 를 통해 가져올 수 있어요.
https://developer.android.com/topic/libraries/architecture/viewmodel-savedstate?hl=ko 참고해보시면 좋을 것 같아요.

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/viewmodel/NewGroupViewModel.kt:
```diff
@@ -0,0 +1,44 @@
+    private val _imageFile = MutableLiveData(File(EMPTY_STRING))
+    val imageFile = _imageFile
+
+    fun postFileUpload(token: String) {
```
p3 : 이름에 `Image` 정보가 들어갔으면 좋겠어요.

ckrudals commented on presentation/src/main/res/layout/activity_new_group.xml:
정렬한번 해주세요

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/viewmodel/KakaoTestViewModel.kt:
```diff
@@ -0,0 +1,37 @@
+
+
+@HiltViewModel
+class KakaoTestViewModel @Inject constructor(
```
p2 : 최신 develop brach 를 pull 하시면 제가만든 `LoginViewModel` 이 있는데요, 이를 활용하면 이 viewModel 은 불필요 할 것 같아요.

ckrudals commented on presentation/src/main/java/com/yapp/bol/presentation/view/login/NewGroupActivity.kt:
```diff
@@ -0,0 +1,115 @@
+
+    private fun getRealPathFromURI(uri: Uri): String {
+        val buildName = Build.MANUFACTURER
+        if (buildName == "Xiaomi") return uri.path ?: EMPTY_STRING
```
p3 : 요 검증은 어떤거에요 ?

prk4224 committed:
[BOL-46] 기능 구현 전 필요한 데이터 추가 및 불필요한 데이터 삭제 (#30)

* refactor:  사용하지 않는 파일 제거

* refactor: theme 변경

* refactor: 오타 수정

* feat: 화면 구현에 필요한 String, color, drawable 추가

* [BOL-46] 이미지 & 프로필 설정 다이얼 로그 구현 (#31)

* feat: 이미지 설정 다이얼 로그 구현

* feat: 프로필 설정 다이얼 로그 구현

* Feat: Mock Prefix 제거

* [BOL-46] 그룹 생성 API 연결 및 기능 구현 (#32)

* feat: GroupApi 생성

* feat: postCreateGroup 기능 구현

* feat: 기능 구현에 필요한 Util 함수 구현

* feat: GalleryManager 생성

* feat: NewGroup 화면 구현

* refactor: safeApiCall 재구성

* [BOL-46] 그룹 생성 완료 기능 구현 (#33)

* feat: NewGroupCompleteActivity 구현

* refactor: Code Clean Up

* refactor: Mock prefix 제거 및 Code Clean Up

* refactor: 컨벤션에 맞게 파일 이름 수정

* [BOL-46] merge conflict 해결  (#35)

* [BOL-15] 네이버 로그인 구현 (#11)

* BOL-15 feat: utils package 에 함수 추가

- flow util 추가
- toast utils 추가

* BOL-15 chore: memory heap size 변경

* BOL-15 style : ktlint code convection & 사용하지 않는 파일 제거

* [BOL-15] feat/ naver 로그인 구현

* BOL-15 chore: yml 에 naver key 추가

* BOL-15 fix : 잘못된 string 값으로 ci 에러나는 이슈 해결

* BOL-15 refactor : presentation 과 data layer 의 파라미터를 관심사에 맡게 분리

* [BOL-11] 구글 로그인  (#12)

* feat: [BOL-11] google login json 파일 추가

* chore: [BOL-11] Firebase 의존성 추가

* style: [BOL-11] ktlint 스타일 수정

* chore: [BOL-11] google login 관련 의존성 추가

* feat: [BOL-11] google login key BuildConfig 추가

* chore: [BOL-11] google-services.json 파일 교체

* chore: [BOL-11] google login app 모듈에 plugin 의존성 추가

* feat: [BOL-11] google login test 용도의 ui 작업

* feat: [BOL-11] google login test 용도의 내부 로직 구현

* BOL-11 refactor: google login lateinit 사용하지 않도록 수정

* feat: [BOL-11] google login json 파일 추가

* chore: [BOL-11] Firebase 의존성 추가

* style: [BOL-11] ktlint 스타일 수정

* chore: [BOL-11] google login 관련 의존성 추가

* chore: [BOL-11] google-services.json 파일 교체

* chore: [BOL-11] google login app 모듈에 plugin 의존성 추가

* feat: [BOL-11] google login test 용도의 ui 작업

* feat: [BOL-11] google login test 용도의 내부 로직 구현

* BOL-11 refactor: google login lateinit 사용하지 않도록 수정

* BOL-11 refactor: 사용하지 않는 import 제거

* BOL-11 feat: ci script에 google key 추가

* BOL-11 fix: google-services 찾지 못하는 에러 수정

* BOL-11 chore: 사용하지 않는 dependency 제거

* BOL-11 refactor: 변수명 구체화 auth -> firebaseAuth

* BOL-11 refactor: 사용하지 않는 전역변수들 지역변수로 변경

* BOL-11 refactor: 사용하지 않는 firebase logic 제거

* BOL-11 chore: firebase dependency 제거

* BOL-11 refactor: 구글 로그인 try-catch 구문 내에서 require로 전체 널체크하도록 refactoring

* BOL-11 refactor: 구글 로그인 버튼 초기화하는 부분 함수명 변경

* BOL-91 fix: 리팩터링한 메소드 변경되지 않아서 발생하는 오류 수정 (#15)

* BOL-46 refactor: Conflict Solve

* [BOL-54]  게임 선택 화면 구현 & 네비게이션 적용 (#36)

* feat: navigation 빌드 추가

* feat: ImageView setImage Util 함수로 분리

* feat: Match Activity  & ViewModel 생성 및 navigation setting

* feat: 그룹 생성 완료 버튼 클릭시 MatchActivity 이동 로직 구현

* feat:  "게임 결과 기록" 상수 추가

* feat: GameItem data class 추가

* feat: 게임 선택 화면 구현

* feat: 멤버 선태 화면 생성

* feat: navigation 에  fragment 추가

* refactor: checked 함수 is 로 변경

* BOL-54  refactor: 사용하지 않는 data class  제거 및 ktlint

* refactor: isEmptyValue -> isInputTextValid 변수 이름 변경

* refactoer: 함수 이름 변경 setImage -> loadImage 및 코드 클린 업

* refactor: MatchViewModel에 HiltViewModel 추가

* refactor: 이름 변경 recyclerAdapter -> gameSelectAdapter

* refactor: 코드 클린 업

* refactor: 불필요한 라인 제거

---------

Co-authored-by: 차경민 <68880067+ckrudals@users.noreply.github.com>
Co-authored-by: EunJung Jung <100047095+eunjjungg@users.noreply.github.com>

---------

Co-authored-by: 차경민 <68880067+ckrudals@users.noreply.github.com>
Co-authored-by: EunJung Jung <100047095+eunjjungg@users.noreply.github.com>

---------

Co-authored-by: 차경민 <68880067+ckrudals@users.noreply.github.com>
Co-authored-by: EunJung Jung <100047095+eunjjungg@users.noreply.github.com>

---------

Co-authored-by: 차경민 <68880067+ckrudals@users.noreply.github.com>
Co-authored-by: EunJung Jung <100047095+eunjjungg@users.noreply.github.com>

---------

Co-authored-by: 차경민 <68880067+ckrudals@users.noreply.github.com>
Co-authored-by: EunJung Jung <100047095+eunjjungg@users.noreply.github.com>

prk4224 committed:
Merge branch 'feat/BOL-46-base' into feat/BOL-46-1

# Conflicts:
#	app/src/main/java/com/yapp/bol/app/di/DataSourceModule.kt
#	buildSrc/src/main/java/com/yapp/bol/Dependencies.kt
#	data/src/main/java/com/yapp/bol/data/mapper/MapperToDomain.kt
#	data/src/main/java/com/yapp/bol/data/model/OAuthApiRequest.kt
#	data/src/main/java/com/yapp/bol/data/repository/MockRepositoryImpl.kt
#	domain/src/main/java/com/yapp/bol/domain/repository/MockRepository.kt
#	domain/src/main/java/com/yapp/bol/domain/usecase/login/LoginUseCase.kt
#	presentation/build.gradle.kts
#	presentation/src/main/java/com/yapp/bol/presentation/view/login/KakaoTestActivity.kt
#	presentation/src/main/java/com/yapp/bol/presentation/viewmodel/MainViewModel.kt
#	presentation/src/main/java/com/yapp/bol/presentation/viewmodel/login/LoginViewModel.kt

prk4224 committed:
refactor: MockDataSource 제거
